packages/react/src/useEditor.ts

```diff
@@ -119,9 +118,9 @@ export const useEditor = (options: Partial<EditorOptions> = {}, deps: Dependency
 
   useEffect(() => {
     return () => {
-      editor?.destroy()
+      return editorRef.current?.destroy()
     }
-  }, [editor])
+  }, [])
```

Comment:
@svenadlung Won't this new logic lead to memory leaks (and potentially other negative side effects)? As of this PR, whenever `deps` change, a `new Editor()` is created. However, this code now only `destroy`s the `editor` on _final_ unmount. So it will _not_ be calling `destroy` on any intermediary `editor`s that are created as `deps` change.

Maybe this can be fixed by moving the `destroy()` call to within the cleanup callback for the `useEffect` above that creates the `editor`. Something like

```ts
  useEffect(() => {
    let isMounted = true

    editorRef.current = new Editor(options)

    editorRef.current.on('transaction', () => {
      requestAnimationFrame(() => {
        requestAnimationFrame(() => {
          if (isMounted) {
            forceUpdate({})
          }
        })
      })
    })
    return () => {
      isMounted = false
      editorRef.current?.destroy();
    }
  }, deps)
```

Similarly, using `editorRef.current` as a dependency for `useEffect` for changing event handlers will not work (see explanation here https://stackoverflow.com/a/60476525/4543977). This means that if the editor changes as `deps` change, it will seemingly _not_ be updating any of the callbacks on the new editor instance. So each new `editor` instance will not be set up correctly.

---

I see the comment in the linked issue https://github.com/ueberdosis/tiptap/issues/4482#issuecomment-1820825154 referencing https://github.com/ueberdosis/tiptap/commit/6984ea11d79c059c55f045038e7758a34872a3f2 as being problematic. Seems that "Destroy editor in safe" PR https://github.com/ueberdosis/tiptap/pull/4000 was potentially the incorrect solution to its bug, and I think similarly this is a bandaid to attempt to fix a new bug that PR introduced, but unfortunately will probably introduce new bugs itself. 😕

Reply:
Good eye @sjdemartini. I'm not sure exactly what `.destroy` entails but on the surface this concern appears to be valid.

On the other hand, it seems possible any loose references will be dropped as they become inaccessible via any code paths - but I think for event listeners this could still be a problem, right?

As far as I know `destroy` only affects the view layer, so is it possible/safe to destroy the old editor before creating the new one? Or would that obliterate history and whatnot? It seems like it would disrupt any decoration state at the very least.

Reply:
At the very least, the new merged implementation will be buggy in that that the event-handlers will *not* be reassigned whenever `deps` change for `useEditor`. So new editor instances will not function properly as dependencies change, which is most certainly a bug. 

But I think it's likely the negative effects will be worse than that, since `destroy` isn't called (event listeners not unassigned from stale versions of the `Editor`, etc.). 

While this new release may fix the flickering with the collaboration cursor, I think that bug should be resolved by going back to what introduced it. It's probably worth looking at the implementation of `useEditor` prior to tiptap 2.1.0, where the flicker was not present.

Reply:
I haven't tried `2.2.x` yet due to other blockers, but I certainly hope that memory leaks won't be introduced to accommodate features we don't use.

It stands to reason that the implementation leaks, but have you done any testing to try to determine if there are indeed memory leaks @sjdemartini?

Reply:
I haven't tested this yet and unfortunately am unlikely to have time, as I'm swamped right now. I'm planning to stay on `v2.0.*` for now, based on the new issues (flickering with the collaboration extension in 2.1, and the new problems here around listeners and/or memory problems in 2.2).

Reply:
Good catch.

For reference, there won't be any memory leak if you keep the dependency array empty.

I would remove the `deps` array altogether because I don't think anyone would intentionally want to re-create the editor.

Reply:
Sure, but that's my whole point: if you actually _do_ need to use the dependency array, this new code will not work properly. The dependency array exists for a reason and has since the early days of Tiptap, as there are legitimate use-cases that require recreating the editor when dependencies change (vs just using methods on the editor itself or something). For instance, you may need to do so if your list of `extensions` change.

Reply:
I believe your assumption has been right @sjdemartini. I've noticed a breaking change going from 2.2.1 (prior to this change) and 2.2.2 (the release of this change). If you call `editor.commands.focus()` from the `onCreate` handler it will work in 2.2.1 and not in 2.2.2. Likely because the handlers haven't been reassigned after a new editor instance was created.

Here's two codesandboxes comparing the two versions:
- [Codesandbox version 2.2.1](https://codesandbox.io/p/sandbox/tiptap-focus-2-2-1-lkwf7q?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clx31xi2u00063b6obsumrcpj%2522%252C%2522sizes%2522%253A%255B100%252C0%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clx31xi2u00023b6o254scxj8%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clx31xi2u00033b6ok0my2skv%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clx31xi2u00053b6opxra42xv%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clx31xi2u00023b6o254scxj8%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clx31xi2t00013b6o6l4cj5jg%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fsrc%252Findex.js%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522id%2522%253A%2522clx31yfk6003d3b6odub97h9l%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A6%252C%2522startColumn%2522%253A27%252C%2522endLineNumber%2522%253A6%252C%2522endColumn%2522%253A27%257D%255D%252C%2522filepath%2522%253A%2522%252Fpackage.json%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522id%2522%253A%2522clx321yat00023b6o66818vxb%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A23%252C%2522startColumn%2522%253A10%252C%2522endLineNumber%2522%253A23%252C%2522endColumn%2522%253A10%257D%255D%252C%2522filepath%2522%253A%2522%252Fsrc%252FApp.js%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522id%2522%253A%2522clx31xi2u00023b6o254scxj8%2522%252C%2522activeTabId%2522%253A%2522clx321yat00023b6o66818vxb%2522%257D%252C%2522clx31xi2u00053b6opxra42xv%2522%253A%257B%2522id%2522%253A%2522clx31xi2u00053b6opxra42xv%2522%252C%2522activeTabId%2522%253A%2522clx31xi2u00043b6otk1hijrt%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clx31xi2u00043b6otk1hijrt%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A0%252C%2522path%2522%253A%2522%252F%2522%257D%252C%257B%2522type%2522%253A%2522SANDBOX_INFO%2522%252C%2522id%2522%253A%2522clx323jds00493b6o8gkxpy11%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%255D%257D%252C%2522clx31xi2u00033b6ok0my2skv%2522%253A%257B%2522tabs%2522%253A%255B%255D%252C%2522id%2522%253A%2522clx31xi2u00033b6ok0my2skv%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Afalse%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)
- [Codesandbox version 2.2.2](https://codesandbox.io/p/sandbox/tiptap-focus-2-2-1-forked-cj88z4?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clx31xi2u00063b6obsumrcpj%2522%252C%2522sizes%2522%253A%255B100%252C0%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clx31xi2u00023b6o254scxj8%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clx31xi2u00033b6ok0my2skv%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clx31xi2u00053b6opxra42xv%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clx31xi2u00023b6o254scxj8%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clx31xi2t00013b6o6l4cj5jg%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fsrc%252Findex.js%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522id%2522%253A%2522clx31yfk6003d3b6odub97h9l%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A6%252C%2522startColumn%2522%253A27%252C%2522endLineNumber%2522%253A6%252C%2522endColumn%2522%253A27%257D%255D%252C%2522filepath%2522%253A%2522%252Fpackage.json%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522id%2522%253A%2522clx321yat00023b6o66818vxb%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A23%252C%2522startColumn%2522%253A10%252C%2522endLineNumber%2522%253A23%252C%2522endColumn%2522%253A10%257D%255D%252C%2522filepath%2522%253A%2522%252Fsrc%252FApp.js%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522id%2522%253A%2522clx31xi2u00023b6o254scxj8%2522%252C%2522activeTabId%2522%253A%2522clx321yat00023b6o66818vxb%2522%257D%252C%2522clx31xi2u00053b6opxra42xv%2522%253A%257B%2522id%2522%253A%2522clx31xi2u00053b6opxra42xv%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clx31xi2u00043b6otk1hijrt%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A0%252C%2522path%2522%253A%2522%252F%2522%257D%255D%252C%2522activeTabId%2522%253A%2522clx31xi2u00043b6otk1hijrt%2522%257D%252C%2522clx31xi2u00033b6ok0my2skv%2522%253A%257B%2522tabs%2522%253A%255B%255D%252C%2522id%2522%253A%2522clx31xi2u00033b6ok0my2skv%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Afalse%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)

Reply:
@samvantoever can you check whether adding a `setTimeout` around your focus call resolves the issue? If so, I would think that suggests race condition rather than handler assignment issue.